pkg/liqo-controller-manager/webhooks/shadowpod/webhook.go

```diff
@@ -0,0 +1,273 @@
+
+	err = peeringInfo.testAndUpdateCreation(ctx, spv.client, shadowpod, *req.DryRun)
+	if err != nil {
+		klog.Warning(err)
+		return admission.Denied(err.Error())
+	}
+
+	return admission.Allowed("")
+}
+
+// HandleUpdate is the function in charge of handling Update requests.
+func (spv *Validator) HandleUpdate(ctx context.Context, req *admission.Request) admission.Response {
+	shadowpod, err := spv.DecodeShadowPod(req.Object)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed decoding of ShadowPod: %w", err))
+	}
+
+	oldShadowpod, err := spv.DecodeShadowPod(req.OldObject)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed decoding of ShadowPod: %w", err))
+	}
+
+	// Check existence and get shadow pod origin Cluster ID label
+	clusterID, found := shadowpod.Labels[forge.LiqoOriginClusterIDKey]
+	if !found {
+		klog.Warningf("Missing origin Cluster ID label on ShadowPod %q", shadowpod.Name)
+		return admission.Denied("missing origin Cluster ID label")
+	}
+
+	// Check existence and get old shadow pod origin Cluster ID label
+	oldClusterID, found := oldShadowpod.Labels[forge.LiqoOriginClusterIDKey]
+	if !found {
+		klog.Warningf("Missing origin Cluster ID label on ShadowPod %q", oldShadowpod.Name)
+		return admission.Denied("missing origin Cluster ID label")
+	}
+
+	if clusterID != oldClusterID {
+		klog.Warningf("The Cluster ID label of the updated shadowpod %s is different from the old one %s", clusterID, oldClusterID)
+		return admission.Denied("shadopow Cluster ID label is changed")
+	}
+
+	if pod.CheckShadowPodUpdate(&shadowpod.Spec.Pod, &oldShadowpod.Spec.Pod) {
+		return admission.Allowed("")
+	}
+
+	return admission.Denied("")
+}
+
+// HandleDelete is the function in charge of handling Deletion requests.
+func (spv *Validator) HandleDelete(ctx context.Context, req *admission.Request) admission.Response {
+	shadowpod, err := spv.DecodeShadowPod(req.OldObject)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed decoding of ShadowPod: %w", err))
+	}
+
+	// Check existence and get shadow pod origin Cluster ID label
+	clusterID, found := shadowpod.Labels[forge.LiqoOriginClusterIDKey]
+	if !found {
+		klog.Warningf("Missing origin Cluster ID label on ShadowPod %q", shadowpod.Name)
+		return admission.Allowed("missing origin Cluster ID label")
+	}
+
+	klog.V(5).Infof("ShadowPod %s decoded: UID: %s - clusterID %s", shadowpod.Name, shadowpod.GetUID(), clusterID)
+	if !spv.enableResourceValidation {
+		return admission.Allowed("")
+	}
+
+	// Check existence and get resource offer by Cluster ID label
+	resourceoffer, err := liqogetters.GetResourceOfferByLabel(ctx, spv.client, corev1.NamespaceAll,
+		liqolabels.LocalLabelSelectorForCluster(clusterID))
+	if err != nil {
+		newErr := fmt.Errorf("error getting resource offer by label: %w", err)
+		klog.Error(newErr)
+		return admission.Allowed(newErr.Error())
+	}
```

Comment:
Should not prevent deletion in case this is not found. I would personally drop completely this, and just retrieve the peeringInfo from the cache (without updating the resources),